### PR TITLE
get command support for json output

### PR DESCRIPTION
**Pull request by rlmcpherson**
This adds support for json output, mirroring the yaml output functionality. Federation, clusters, and instancegroups are supported. Secrets are not supported, as in the yaml code. Includes a small refactoring of the get formatting code for conciseness, with a single function called to write out both json and yaml.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1155)

### CODE REVIEW

File: cmd/kops/get.go [hunk 1]
```diff
@@ -39,12 +45,45 @@ var getCmd = GetCmd{
...
+	y, err := api.ToVersionedYaml(obj)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling yaml: %v", err)
+	}
+	return y, nil
+}
+
+// v must be a pointer to a marshalable object
+func marshalJSON(obj runtime.Object) ([]byte, error) {
+	j, err := json.MarshalIndent(obj, "", "  ")
```
Comment (justinsb) [inline cmd/kops/get.go]: We probably should build api.ToVersionedJson, but I'm also still learning the API machinery so we can leave this for now :-)

File: cmd/kops/get.go [hunk 2]
```diff
@@ -39,12 +45,45 @@ var getCmd = GetCmd{
...
 
 	rootCommand.AddCommand(cmd)
 
-	cmd.PersistentFlags().StringVarP(&getCmd.output, "output", "o", OutputTable, "output format.  One of: table, yaml")
+	cmd.PersistentFlags().StringVarP(&getCmd.output, "output", "o", OutputTable, "output format.  One of: table, yaml, json")
+}
+
+type marshalFunc func(obj runtime.Object) ([]byte, error)
+
+func marshalToStdout(obj runtime.Object, marshal marshalFunc) error {
```
Comment (justinsb) [inline cmd/kops/get.go]: Probably better to pass the io.Writer in here, instead of assuming os.Stdout

File: cmd/kops/get_cluster.go [hunk 3]
```diff
@@ -107,36 +109,57 @@ func (c *GetClustersCmd) Run(args []string) error {
...
-			y, err := api.ToVersionedYaml(cluster)
-			if err != nil {
-				return fmt.Errorf("error marshaling yaml for %q: %v", cluster.Name, err)
+			if err := marshalToStdout(cluster, marshalYaml); err != nil {
+				return err
 			}
-			_, err = os.Stdout.Write(y)
+		}
+		return nil
+	case OutputJSON:
```
Comment (justinsb) [inline cmd/kops/get_cluster.go]: Probably cleaner to do (pseudocode):

```
case OutputJSON, OutputYaml:
  f := marshalYaml
  if mode == OutputJSON {
    f = marshalJSON
  }

...
```

to avoid the code repetition (though I see you've got rid of most of it with fullClusterSpecs.

File: cmd/kops/get_federation.go [hunk 4]
```diff
@@ -82,19 +84,21 @@ func RunGetFederations(context Factory, out io.Writer, options *GetFederationOpt
 			return strings.Join(f.Spec.Members, ",")
 		})
 		return t.Render(federations, out, "NAME", "CONTROLLERS", "MEMBERS")
-	} else if output == OutputYaml {
+
+	case OutputYaml:
```
Comment (justinsb) [inline cmd/kops/get_federation.go]: Oh ... idea: We could declare a function that maps OutputYaml -> marshalYaml etc.  We could even make OutputYaml a named type (`type OutputFormat string`) and then declare a function on `OutputFormat` (I think)

File: cmd/kops/get.go (hunk 2 above)
Comment (rlmcpherson) [inline cmd/kops/get.go]: Good point, I'll update.

File: cmd/kops/get.go (hunk 1 above)
Comment (rlmcpherson) [inline cmd/kops/get.go]: Yes, I started adding this before updating and seeing the change to the yaml function to include versioning. Also, since the api object structs are already decorated with json tags it doesn't seem strictly necessary for marshaling json.

File: cmd/kops/get_cluster.go (hunk 3 above)
Comment (rlmcpherson) [inline cmd/kops/get_cluster.go]: That seemed to work fine and eliminates the repetition. Pushed.

File: cmd/kops/get_federation.go (hunk 4 above)
Comment (rlmcpherson) [inline cmd/kops/get_federation.go]: I think that could work. What about the case of OutputTable, since the code differs substantially between cluster, isntancegroup, and federation. Maybe I'm misunderstanding? Either way, we can iterate on this as you said.

**Comment by rlmcpherson** [comment]
@justinsb Updated per the comments. Anything else to address before merging?

File: cmd/kops/get_federation.go (hunk 4 above)
Comment (justinsb) [inline cmd/kops/get_federation.go]: I agree that OutputTable should be separate.  At least it isn't obvious that it should be the same, so that could always be a separate PR.  Sorry not to be clear!
